File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/ArtistLibraryIterator.cs
namespace IteratorDesignPattern.Example3;

public class ArtistLibraryIterator : IIterator<Song>
{
    private readonly ArtistLibrary _library;
    private readonly List<Song> _allSongs;
    private int _currentIndex = 0;

    public ArtistLibraryIterator(ArtistLibrary library)
    {
        _library = library;
        _allSongs = new List<Song>();
        
        foreach (var artistSongs in _library.GetArtistSongs().Values)
        {
            _allSongs.AddRange(artistSongs);
        }
    }

    public bool HasNext()
    {
        return _currentIndex < _allSongs.Count;
    }

    public Song Next()
    {
        return _allSongs[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/ArtistLibraryIterator.java
package behavioral.iterator.example3;

import java.util.*;

public class ArtistLibraryIterator implements IIterator<Song> {
    private final List<Song> allSongs = new ArrayList<>();
    private int currentIndex = 0;

    public ArtistLibraryIterator(ArtistLibrary library) {
        for (List<Song> songs : library.getArtistSongs().values()) {
            allSongs.addAll(songs);
        }
    }

    @Override
    public boolean hasNext() { return currentIndex < allSongs.size(); }

    @Override
    public Song next() { return allSongs.get(currentIndex++); }

    @Override
    public void reset() { currentIndex = 0; }
}
